working/0723-static-extensions/feature-specification-variant1.md

```diff
@@ -0,0 +1,467 @@
+`X1 .. Xs` such that the instantiated constructor return type is
+exactly `C<U1 .. Uk>`, it is not sufficient that it is a subtype thereof,
+or that it differs in any other way.*
+
+A compile-time error occurs if the invocation passes actual type arguments
+to both `E` and `C`, call them `S1 .. Ss` and `U1 .. Uk`, respectively,
+unless the instantiated constructor return type of _D_ with actual type
+arguments `S1 .. Ss` is `C<U1 .. Uk>`. In this type comparison, top types
+like `dynamic` and `Object?` are considered different, and no type
+normalization occurs. *In other words, the types must be equal, not
+just mutual subtypes.*
+
+*Note that explicitly resolved invocations of constructors declared in
+static extensions are an exception in real code, usable in the case where a
+name clash prevents an implicitly resolved invocation. However, implicitly
+resolved invocations are specified in the rest of this section by reducing
+them to explicitly resolved ones.*
+
+A constructor invocation of the form `C<T1 .. Tm>.name(args)` is partially
+resolved by looking up a constructor named `C.name` in the class `C` and in
+every accessible static extension with on-class `C`. A compile-time error
+occurs if no such constructor is found. Similarly, an invocation of the
+form `C<T1 ... Tm>(args)` uses a lookup for constructors named `C`.
+
+If a constructor in `C` with the requested name was found, the pre-feature
+static analysis and dynamic semantics apply. *That is, the class always
+wins.*
+
+Otherwise, the invocation is partially resolved to a set of candidate
+constructors found in static extensions. Each of the candidates _kj_ is
+vetted as follows:
+
+Assume that _kj_ is a constructor declared by a static extension _D_ named
+`E` with type parameters `X1 extends B1 .. Xs extends Bs` and on-type
+`C<S1 .. Sm>`.  Find actual values `U1 .. Us` for `X1 .. Xs` satisfying the
+bounds `B1 .. Bs`, such that `([U1/X1 .. Us/Xs]C<S1 .. Sm>) == C<T1 .. Tm>`.
+If this fails then remove _kj_ from the set of candidate constructors.
+Otherwise note that _kj_ uses actual type arguments `U1 .. Us`.
+
+If all candidate constructors have been removed, or more than one candidate
+remains, a compile-time error occurs. Otherwise, the invocation is
+henceforth treated as `E<U1 .. Us>.C<T1 .. Tm>.name(args)` (respectively
+`E<U1 .. Us>.C<T1 .. Tm>(args)`).
+
+A constructor invocation of the form `C.name(args)` (respectively
+`C(args)`) where `C` denotes a non-generic class is resolved in the
+same manner, with `m == 0`. *In this case, type parameters declared by `E`
+will be bound to values selected by instantiation to bound.*
+
+Consider a constructor invocation of the form `C.name(args)` (and similarly
+for `C(args)`) where `C` denotes a generic class. As usual, the
+invocation is treated as in the pre-feature language when it denotes a
+constructor declared by the class `C`.
+
+In the case where the context type schema for this invocation fully
```

Comment:
This all feels like it needs some work.  I'm not sure if I have a full model of this in my head, but my rough initial cut is something like the following (by example).  If we have an extension declaration as follows:
```dart
extension E<S, T> on Map<int, T> {
  Map.bar(S, x, T y) => {3 : y};
}
```
we should semantically view this as if we had declared a static method something like the following:
```dart
  static Map<int, T> E_Map_bar<S, T>(S x, T y) => E<S, T>.Map.bar(x);
```
and hence when we encounter something like `Map.bar("hello")` in downwards context `K`, we treat it as an inference problem equivalent to solving `E_Map_bar("hello")` in downwards context `K`.  What do we do if there are explicit type arguments passed though?  That is, what is the interpretation of `Map<T1, T2>.bar("hello")` in downwards context `K`?  I think one possible answer is that we treat this as equivalent to solving `E_Map_bar("hello")` in downwards context `Map<T1, T2>`.  I'm not 100% sure about that though.  WDYT?

Reply:
Filed https://github.com/dart-lang/language/issues/4053 to discuss this.

Reply:
Agreed, this needs some work. ;-) I'll use 4053 to improve on it later. Let me try to see if I can respond to the case at this time as well:

```dart
extension E<S, T> on Map<int, T> {
  Map.bar(S x, T y) => {3 : y};
}

void main() {
  K map = Map.bar(1.5, "hello");
}
```

I think we need to resolve at first that `Map.bar` means `E.Map.bar`. There are different ways to do this, but I've aimed to specify that we accept the declaration in the class `Map` if it exists ("class always wins"), otherwise we find all static extensions declaring a constructor `Map.bar` and/or a static member of `Map` named `bar`. If there are both static members and constructors then it's an error. Assume that we have only constructors. If there are two or more static extensions declaring a constructor named `Map.dar` then it's an error (because we have no idea what the actual type arguments would be, and then we don't want to try to filter the extensions).

OK, so it's `E.Map.bar(1.5, "hello")`.

There will be some actual type arguments of `Map`, and the type of the expression will be `Map<...>` with those actual type arguments.

So we use the downward context `K` on the on-type of `E` (not on the entire signature, it is only the constructor return type that matters):

```dart
Map<int, T> E_Map_bar<S, T>(S x, T y) => throw "For inference only";

void main() {
  K map = E_Map_bar(1.5, "Hello");
}
```

This will bind `S` to `double` (there's no way `K` can influence that), and `T` must be some supertype of `String`, and we will allow the downward phase to choose the type argument "if `K` has an opinion" about that (`K == dynamic` doesn't, but `K == Map<Object, Object>` does).